builder/Builder.java

```diff
@@ -0,0 +1,382 @@
+        input.transferTo(output);
+        return Optional.of(output.toString("UTF-8"));
+      } catch (Exception exception) {
+        // TODO debug("read({0}) failed: {1}", uri, exception);
+        return Optional.empty();
+      }
+    }
+
+    /** Load content from specified uri into temporary file. */
+    static Optional<Path> load(URI uri) {
+      try {
+        URL url = uri.toURL();
+        String[] parts = url.getFile().split("/");
+        Path temp = Files.createTempFile("", "-" + parts[parts.length - 1]);
+        try (InputStream sourceStream = url.openStream();
+            OutputStream targetStream = Files.newOutputStream(temp)) {
+          sourceStream.transferTo(targetStream);
+        } catch (Exception e) {
+          Files.deleteIfExists(temp);
+          return Optional.empty();
+        }
+        return Optional.of(temp);
+      } catch (Exception e) {
+        return Optional.empty();
+      }
+    }
+
+    static Optional<String> scan(URI uri) {
+      // TODO debug("scanJar({0} -> {1})", uri, value);
+      Optional<Path> path = Util.load(uri);
+      if (!path.isPresent()) {
+        // TODO System.out.printf("scanJar(%s) failed%n", uri);
+        return Optional.empty();
+      }
+      try {
+        return scan(path.get(), true);
+      } finally {
+        try {
+          Files.delete(path.get());
+        } catch (IOException e) {
+          // TODO System.err.println("deleting temp file failed: " + e);
+        }
+      }
+    }
+
+    static Optional<String> scan(Path path, boolean reportFileNameBasedModuleAsEmpty) {
+      try {
+        Set<ModuleReference> allModules = ModuleFinder.of(path).findAll();
+        if (allModules.size() != 1) {
+          throw new IllegalStateException("expected to find single module, but got: " + allModules);
+        }
+        ModuleReference reference = allModules.iterator().next();
+        ModuleDescriptor descriptor = reference.descriptor();
+        String module = descriptor.name();
+        System.out.println(descriptor + " packaged in " + path);
+        if (reportFileNameBasedModuleAsEmpty) {
+          if (descriptor.isAutomatic()) {
+            if (!isAutomaticModuleNameAttributeAvailable(reference)) {
+              return Optional.empty();
+            }
+          }
+        }
+        return Optional.of(module);
+      } catch (FindException e) {
+        // TODO debug("scan failed: {0}", e);
+        return Optional.empty();
+      }
+    }
+
+    static boolean isAutomaticModuleNameAttributeAvailable(ModuleReference moduleReference) {
+      try (ModuleReader moduleReader = moduleReference.open()) {
+        String manifestString =
+            moduleReader
+                .read("META-INF/MANIFEST.MF")
+                .map(StandardCharsets.UTF_8::decode)
+                .map(Object::toString)
+                .orElse("");
+        if (manifestString.contains("Automatic-Module-Name")) {
```

Comment:
Will need to check for `module-info.class` too, unless I missed it.

Reply:
That check is _a priori_ done by the `ModuleFinder` usage. Only "automatic" modules are required to be stable, i.e. define an `Automatic-Module-Name` attribute.

Reply:
I'll add a flag to the table that indicates whether a module is explicit or autotmatic.